In [469]:
# Import required libraries and dependencies
import pandas as pd
import numpy as np
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [471]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [473]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [475]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [479]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

# Fit the scaler to the data and transform it
scaled_data = scaler.fit_transform(df_market_data)

In [481]:
# Create a DataFrame with the scaled data
scaled_df = pd.DataFrame(
    scaled_data,
    columns=df_market_data.columns,
    index=df_market_data.index
)

# Copy the crypto names from the original DataFrame
if 'coin_id' in df_market_data.columns:
    scaled_df['coin_id'] = df_market_data['coin_id']
  
# Set the coin_id column as index
    scaled_df.set_index('coin_id', inplace=True)
else:
    scaled_df.index = df_market_data.index

# Display the scaled DataFrame
display(scaled_df.head())

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [485]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [487]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in range(1, 12):
    
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=42, n_init='auto')
    
    # 2. Fit the model to the data using scaled_data
    kmeans.fit(scaled_df)

    # 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans.inertia_)

# Print the results
print("Inertia values for k=1 to 11:")
for k, i in zip(range(1, 12), inertia):
    print(f"k={k}: Inertia = {i:.2f}")
    

Inertia values for k=1 to 11:
k=1: Inertia = 287.00
k=2: Inertia = 212.12
k=3: Inertia = 170.19
k=4: Inertia = 79.02
k=5: Inertia = 71.10
k=6: Inertia = 66.11
k=7: Inertia = 54.49
k=8: Inertia = 43.22
k=9: Inertia = 32.86
k=10: Inertia = 28.40
k=11: Inertia = 24.12


In [489]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': list(range(1,12)),
    'inertia': inertia
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

# Display the DataFrame
display(elbow_df)


k     inertia
0    1  287.000000
1    2  212.123342
2    3  170.193571
3    4   79.022435
4    5   71.097914
5    6   66.108827
6    7   54.488123
7    8   43.224693
8    9   32.860884
9   10   28.404885
10  11   24.116198

In [491]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_curve = elbow_df.hvplot.line(
   x="k", y="inertia", label="Raw Data", color="blue", line_dash='solid'
)

# Print the plot
elbow_curve

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 6

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [496]:
# Initialize the K-Means model using the best value for k
optimal_k = 6
kmeans = KMeans(n_clusters=optimal_k, random_state=0)

In [498]:
# Fit the K-Means model using the scaled DataFrame
kmeans.fit(scaled_df)

KMeans(n_clusters=6, random_state=0)

In [500]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = kmeans.predict(scaled_df)

# Print the resulting array of cluster values.
print(clusters)

[2 2 0 0 5 2 5 2 2 0 0 0 0 5 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 4 2 0 1 2 0 0 3
 0 2 0 0]


In [502]:
# Create a copy of the scaled DataFrame
clustered_df = scaled_df.copy()

In [504]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
clustered_df['Cluster'] = clusters

# Display the copy of the scaled DataFrame
display(clustered_df.head())

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        5

In [506]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scatter_plot = clustered_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Cluster",  
    hover_cols=["coin_id"],
    title="Cryptocurrency Clusters by 24h vs 7d Price Changes",
    width=800,
    height=500,
    alpha=0.7,
    size=100,   
    legend='top_right',
    colorbar=True,
    cmap='Category10' 
)

# Display the plot
scatter_plot

:NdOverlay   [Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [510]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [512]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(scaled_df)

# View the scaled PCA data
pca_df = pd.DataFrame(
    data=pca_data,
    columns=["PC1", "PC2", "PC3"],
    index=scaled_df.index  
)

# Print the scaled PCA data 
print(pca_df)

                           PC1       PC2       PC3
coin_id                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 

In [514]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
cumulative_variance = np.cumsum(explained_variance)

# Create variance explanation table
variance_df = pd.DataFrame({
    'Principal Component': ["PC1", "PC2", "PC3"],
    'Individual Variance': explained_variance,
    'Cumulative Variance': cumulative_variance
}).set_index('Principal Component')

# Display results
display(display_df)

Individual Variance Cumulative Variance
Principal Component                                        
PC1                              37.20%              37.20%
PC2                              34.70%              71.90%
PC3                              17.60%              89.50%

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 89.50%

In [517]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(
    data=pca_data,
    columns=[f"PC{i+1}" for i in range(pca_data.shape[1])]
)

# Copy the crypto names from the original scaled DataFrame
pca_df['coin_id'] = df_market_data.index.values

# Set the coin_id column as index
pca_df.set_index('coin_id', inplace=True)

# Display the scaled PCA DataFrame
display(pca_df.head())

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [414]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [416]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    
    # Inside the loop:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=42)

    # 2. Fit the model to the data using `pca_df`
    kmeans.fit(pca_df)

    # 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans.inertia_)

In [418]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_optimized = {
    "k": k_values,  
    "inertia": inertia        
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_optimized = pd.DataFrame(elbow_data_optimized)

# Display the DataFrame
display(elbow_data_optimized)

k     inertia
0    1  256.874086
1    2  182.339530
2    3  140.431833
3    4   49.665497
4    5   42.051901
5    6   31.749005
6    7   28.031869
7    8   19.451747
8    9   13.814082
9   10   10.630648
10  11   10.120243

In [420]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot_optimized = elbow_data_optimized.hvplot.line(
    x="k", y="inertia", label="Optimized Data", color="green", line_dash='dashed'
)

# Display the plot
elbow_plot_optimized

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4. This suggests that 4 clusters capture the majority of the data. 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Yes, the original data suggested k=6, while PCA reduced it to k=4. 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [424]:
# Initialize the K-Means model using the best value for k
kmeans_pca = KMeans(n_clusters=4, random_state=42, n_init=10)  

In [426]:
# Fit the K-Means model using the PCA data
kmeans_pca.fit(pca_df)

KMeans(n_clusters=4, n_init=10, random_state=42)

In [428]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
clusters_labels = kmeans_pca.predict(pca_df)

# Print the resulting array of cluster values.
print(clusters_labels)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 3 0 2 2 1
 2 2 2 2]


In [430]:
# Create a copy of the scaled PCA DataFrame
pca_df_clusters = pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
pca_df_clusters['Cluster'] = kmeans_pca.labels_

# Display the copy of the scaled PCA DataFrame
display(pca_df_clusters)

PC1       PC2       PC3  Cluster
coin_id                                                    
bitcoin              -0.600667  0.842760  0.461595        0
ethereum             -0.458261  0.458466  0.952877        0
tether               -0.433070 -0.168126 -0.641752        2
ripple               -0.471835 -0.222660 -0.479053        2
bitcoin-cash         -1.157800  2.041209  1.859715        0
binancecoin          -0.516534  1.388377  0.804071        0
chainlink            -0.450711  0.517699  2.846143        0
cardano              -0.345600  0.729439  1.478013        0
litecoin             -0.649468  0.432165  0.600303        0
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653        2
crypto-com-chain     -0.248198 -1.376252 -1.462026        2
usd-coin             -0.438408 -0.175337 -0.663388        2
eos                  -0.693425 -0.473815 -0.527597        2
monero                0.060499  2.909404  1.498571        0
tron                 -0.393352 -0.108192 -0.012756        2
tezos                -0.796176 -0.494409  1.082812        0
okb                   0.064075 -1.269825 -1.098829        2
stellar              -0.489015 -0.732719 -0.062543        2
cosmos               -0.306272  0.703415  1.714224        0
cdai                 -0.513528 -0.142802 -0.656566        2
neo                  -0.362120 -0.986914 -0.728752        2
wrapped-bitcoin      -0.604265  0.827398  0.439316        0
leo-token            -0.413296 -0.674115 -1.076628        2
huobi-token          -0.407483 -0.212507 -0.351426        2
nem                   0.608974  0.563532 -1.148742        2
binance-usd          -0.450211 -0.151019 -0.647401        2
iota                 -0.764665 -0.517886  0.204990        2
vechain              -0.556315 -1.938209 -1.261776        2
zcash                -0.425147  0.492976  1.058048        0
theta-token           2.676868 -0.013954 -1.965207        2
dash                 -0.613923 -0.479337  0.339565        2
ethereum-classic     -0.579924 -0.356334 -0.114942        2
ethlend               8.089018 -3.896891  2.301382        3
maker                -0.389045  0.165041  0.379414        0
havven                0.865762 -2.261882  0.275583        2
omisego               0.111675  0.428316 -1.205398        2
celsius-degree-token  4.792395  6.767679 -1.986985        1
ontology             -0.632355 -2.108117 -0.652227        2
ftx-token            -0.593142  0.021485  0.209911        2
true-usd             -0.458131 -0.135734 -0.635284        2
digibyte             -0.297910 -0.191126 -0.909602        2

In [445]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# The cryptocurrency represented by each data point.
scatter_plot_optimized = pca_df_clusters.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="Cluster",  
    hover_cols=["coind_id"],
    title="Cryptocurrency Clusters by PCA Components",
    width=800,
    height=500,
    alpha=0.7,
    size=100,   
    legend='top_right',
    colorbar=True,
    cmap='Category10' 
)

# Display the plot
scatter_plot_optimized

:NdOverlay   [Cluster]
   :Scatter   [PC1]   (PC2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [459]:
# Composite plot to contrast the Elbow curves
elbow_curves = (elbow_curve * elbow_plot_optimized).opts(
    title="Elbow Curve Comparison: Raw vs Optimized",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    legend_position="top_right",
    width=700,
    height=400
)

# Display the composite plot
elbow_curves

:Overlay
   .Curve.Raw_Data       :Curve   [k]   (inertia)
   .Curve.Optimized_Data :Curve   [k]   (inertia)

In [526]:
# Composite plot to contrast the clusters

# Define plot1
plot1 = clustered_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Cluster",
    title="24h vs 7d Price Changes",
)

# Define plot2
plot2 = pca_df_clusters.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="Cluster",
    title="PCA Components",
)

# Combine the plots
combined = plot1 + plot2
combined.opts(title="Combined Cryptocurrency Analysis")

:Layout
   .NdOverlay.I  :NdOverlay   [Cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .NdOverlay.II :NdOverlay   [Cluster]
      :Scatter   [PC1]   (PC2)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using PCA Components for the Cryptocurrency Analysis streamlines clustering and reduces noise. In the PCA-reduced scatter plot (PC1 vs. PC2), the clusters appear tighter and closer than in the original plot (24h vs. 7d changes). This suggests that the K-Means model found stable patterns, and this is useful to help identify crypto categories.